### ISSUE TITLE
Needs an option to add MapToStoredProcedures() in the Configuration Classes

### ISSUE THREAD

**Issue by rposener** [ReversePOCO/EntityFramework-Reverse-POCO-Code-First-Generator#117]
There is no option to use MapToStoredProcedures() in the configuration classes.  This should be a standard option.  Instead I'm going to have to use a method to add this during the OnModelCreatingPartial, which isn't horrible, just seems like a waste to cycle back over my 400+ classes to add this.  Our model is ridiculously large already and unfortunately not easily broken up.

**Comment by AndyCadley** [comment]
Yeah, I found this to be a bit of a pain and something that doing it via partial classes doesn't really scale well (unless you add yet more code generation). I've worked around by adding the following to the Core .ttinclude file:

```C#
        public enum Operation
        {
            Insert,
            Update,
            Delete
        };
        bool MapToStoredProcedures = false;
        Func<Table,Operation,string> MapToStoredProceduresRename;
```
in the appropriate places. And then the following to the OnModelCreating generation in the other .ttinclude file (just after the modelBuilder.Configuration.Add line)
```C#
<#if(MapToStoredProcedures) {#>
            modelBuilder.Entity<<#=tbl.NameHumanCase #>>().MapToStoredProcedures(
<#if(MapToStoredProceduresRename != null) {#>
                p => p.Insert(i => i.HasName("<#=MapToStoredProceduresRename(tbl, Operation.Insert)#>"))
                .Update(u => u.HasName("<#=MapToStoredProceduresRename(tbl, Operation.Update)#>"))
                .Delete(d => d.HasName("<#=MapToStoredProceduresRename(tbl, Operation.Delete)#>"))
<# }#>
                );
<# }#>
```

This will then allow you to define the mapping in you main .tt settings file with something like:
```C#
    // Mapping entites to stored procedures
    MapToStoredProcedures = true;
    MapToStoredProceduresRename = (t,o) => {
        return $"{t.Schema}.P_{o.ToString()}_{t.NameHumanCase}";
        };
```
If you leave MapToStoredProceduresRename with null, it'll use the default Entity Framework mappings or you can override it to apply your own preferred naming conventions.

It's not a comprehensive solution (it doesn't really account for column remapping or selective use of stored procedures) but it's a lot easier than manually doing all the work.

**Comment by sjh37** [comment]
Excellent work @AndyCadley I'll get this added.

**Comment by ackginger** [comment]
I'm already using this locally as it is just what we need. Thanks!

**Comment by ackginger** [comment]
Having said that ... I have also had to add a MapToStoredProceduresTableFilterExclude regex, since it turns out our tables are not all using custom stored procs. Would be great to have this granularity and simply required changing @AndyCadley's code above to 

`<#if(MapToStoredProcedures && (MapToStoredProceduresTableFilterExclude == null || !MapToStoredProceduresTableFilterExclude.IsMatch(tbl.Name))) {#>`

**Comment by sjh37** [comment]
EF6 can do this for every entity in one line with a lightweight convention, so there's no need for per-class code or a generator setting. Make the context partial:

```csharp
Settings.DbContextClassModifiers = "public partial";
```

and implement the partial method in your own file:

```csharp
public partial class MyDbContext
{
    partial void OnModelCreatingPartial(DbModelBuilder modelBuilder)
    {
        modelBuilder.Types()
            .Where(t => !t.Name.StartsWith("Legacy")) // optional: exclude the tables that don't use procedures
            .Configure(t => t.MapToStoredProcedures());
    }
}
```

The `.Where(...)` gives you the per-table exclusion asked for above.

Closing, as this is possible with v4.
